### PR TITLE
Improve rowkey object size estimate

### PR DESCRIPTION
**Pull request by adarshsanjeev**
Improves the size estimation of Rowkey by accounting for object overhead. This will improve the accuracy when estimating size of sketches in MSQ.

- Improves the size estimation of Rowkey

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/frame/key/RowKey.java [hunk 1]
```diff
@@ -109,8 +109,12 @@ public String toString()
...
 
-  public int getNumberOfBytes()
+  /**
+   * Estimate number of bytes taken by an object of {@link RowKey}. Only returns an estimate and does not account for
+   * platform or JVM specific implementation.
+   */
+  public int estimatedObjectSize()
   {
-    return array().length;
+    return 44 + array().length; // Constant to account for hashcode and object overhead
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/frame/key/RowKey.java:118]:
```suggestion
    private static final int OBJECT_OVERHEAD_SIZE_BYTES = 44;
    return OBJECT_OVERHEAD_SIZE_BYTES + array().length; // Constant to account for hashcode and object overhead
```

File: processing/src/main/java/org/apache/druid/frame/key/RowKey.java [hunk 2]
```diff
@@ -109,8 +109,12 @@ public String toString()
     return Arrays.toString(key);
   }
 
-  public int getNumberOfBytes()
+  /**
+   * Estimate number of bytes taken by an object of {@link RowKey}. Only returns an estimate and does not account for
+   * platform or JVM specific implementation.
+   */
+  public int estimatedObjectSize()
```
Comment (LakshSingla) [inline processing/src/main/java/org/apache/druid/frame/key/RowKey.java:116]: I think we should also append a memory unit to the method name like `estimatedObjectSizeBytes`

Comment (adarshsanjeev) [reply processing/src/main/java/org/apache/druid/frame/key/RowKey.java:116]: Done

File: processing/src/main/java/org/apache/druid/frame/key/RowKey.java [hunk 3]
```diff
@@ -31,6 +31,7 @@
 public class RowKey
 {
   private static final RowKey EMPTY_KEY = new RowKey(new byte[0]);
+  private static final int OBJECT_OVERHEAD_SIZE_BYTES = 44; // Constant to account for hashcode and object overhead
```
Comment (cryptoe) [inline processing/src/main/java/org/apache/druid/frame/key/RowKey.java:34]: Can we add each item individually or mention in the comment how 44 was arrived at?

**Review by cryptoe — APPROVED** [review]
LGTM. Thanks for the fix.

**Comment by cryptoe** [comment]
Failures look unrelated. Thanks for the fix @adarshsanjeev !!
As part of the next changes, we can try to add the byte[] directly to the sketch which would help us save 44 bytes.

**Pull request merged**